File: lab3/src/chroma_hpcp.m
clear all
close all

filenames = {'melody4', 'ALoDown-mono', '07', 'extrabits'};
%filenames = {'melody4'};
for i=1:size(filenames,2)
    % Read from CSV
    C = csvread(sprintf('hpcp_csv/%s_vamp_vamp-hpcp-mtg_MTG-HPCP_HPCP.csv',filenames{i}))';
    tt = C(1,:);
    clear C2
    C2(1:9,:) = C(5:13,:);
    C2(10:12,:) = C(2:4,:);
    clear C
    C = C2;
    
    % Plot Chromagram
    fig = 1;
    figure(fig)
    imagesc(tt,[1:12],20*log10(C+eps));
    axis xy
    caxis(max(caxis)+[-60 0])
    aux=(1:12); 
    set(gca,'ytick',aux); 
    set(gca,'YTickLabel',{'C';'C#';'D';'D#';'E';'F';'F#';'G';'G#';'A';'A#';'B'});
    title(sprintf('Chromagram | %s.wav',filenames{i}))

    nameplot = sprintf('Chromagram_%s.jpg',filenames{i});
    saveas(figure(fig), nameplot);

    % Plot average chroma features
    fig = fig+1;
    figure(fig) 
    bar(mean(C')); 
    set(gca,'xtick',aux); 
    %    set(gca,'XTickLabel',{'A';'A#';'B';'C';'C#';'D';'D#';'E';'F';'F#';'G';'G#'; });
    set(gca,'XTickLabel',{'C';'C#';'D';'D#';'E';'F';'F#';'G';'G#';'A';'A#';'B'});
    grid
    title(sprintf('Average chroma features | %s.wav',filenames{i}));
    
    nameplot = sprintf('InstantChroma_%s.jpg',filenames{i});
    saveas(figure(fig), nameplot);
end

File: lab3/src/chroma_hpcp.py
# -*- coding: utf-8 -*-
from pylab import *
import os
from distutils.dir_util import mkpath
import csv
#from utils import *
#from descriptors import *   

folder = "hpcp_csv" # directory where CSVs wibe created
#filenames = ('melody4', 'ALoDown-mono', '07', 'extrabits')
filenames = range(1,97) #97
transfile = 'hpcp.n3' # use the same for all melodies   

for filename in filenames:    
    path = '../mirex_key/audio/%02d.wav' % filename # path to audio file
    mkpath(folder) # create dir for ouput. this does not fail if dir already exists
    os.system("rm %s/%02d_vamp_vamp-hpcp-mtg_MTG-HPCP_HPCP.csv" % (folder, filename)) # Remove if csv's already exist
    #os.system("/usr/local/bin/sonic-annotator -s vamp:vamp-hpcp-mtg:MTG-HPCP:HPCP > %s" % (transfile)) # create configuration file
    os.system("/usr/local/bin/sonic-annotator -t %s %s -w csv --csv-basedir %s" % (transfile, path, folder)) # extract features

    # We want to reorder the data so columns are output in the form: time, C, C#, D, ..., B
    data = genfromtxt("%s/%02d_vamp_vamp-hpcp-mtg_MTG-HPCP_HPCP.csv" % (folder, filename), delimiter=',')
    chroma = data.copy()
    chroma[:, 1:data.shape[1]-3] = data[:, 4:data.shape[1]]
    chroma[:, data.shape[1]-3:data.shape[1]] = data[:, 1:4]        # good?
    #savetxt("%s/%02d.csv" % (folder, filename), chroma, delimiter=',', fmt='%10.7f')
    savetxt("%s/%02d_avg.csv" % (folder, filename), mean(chroma[:, 1:],0), delimiter=',') # save key profile    
    
    os.system("rm %s/%02d_vamp_vamp-hpcp-mtg_MTG-HPCP_HPCP.csv" % (folder, filename))
    